### ISSUE TITLE
wrong type check in isFunction

### ISSUE THREAD

**Issue by DanielRuf** [jquery/jquery#3955]
https://github.com/jquery/jquery/commit/a16339b8933f115da3661f3d3f64854c3fa60bdc introduced a bug. The typecheck is not correct. The `toType` method was not used and produces failing unit tests at https://github.com/zurb/foundation-sites

**Comment by gibson042** [comment]
Please provide a link to a reduced test case demonstrating your issue.

**Comment by DanielRuf** [comment]
Basically the failing unit tests at https://github.com/zurb/foundation-sites/commits/develop

**Comment by timmywil** [comment]
A link to unit tests a test case is not. We need something reduced to demonstrate a clear issue in isFunction. There's a lot going on in those tests and one of the failures follows a codepath that never touches jQuery. And how `getAttribute` being undefined on a supposed script tag led you to `isFunction` is entirely unclear.

**Comment by timmywil** [comment]
BTW, jsfiddle.net or jsbin.com or good places to start.

**Comment by timmywil** [comment]
> `git bisect` showed this exact commit and change as I already wrote

Not to be mean, but this doesn't help us at all. We have to require a test case or we'd be forced to sift through everyone's applications. Ain't nobody got time for that.

**Comment by DanielRuf** [comment]
> I'll try to create a reduced testcase when I find the time for it this week.

**Comment by DanielRuf** [comment]
It was supposed to be an answer to
> And how getAttribute being undefined on a supposed script tag led you to isFunction is entirely unclear.

**Comment by timmywil** [comment]
> It was supposed to be an answer

I see what you mean now. git bisect led you there. Fair enough, but what's still unclear is how the two are related. Thank you for when you get to the test case. We'll be happy to take a closer look then.

**Comment by DanielRuf** [comment]
Generally I also miss the changes for https://github.com/jquery/jquery-migrate as this is a big breaking change here.

I am currently revisitng this issue and trying to fix some parts in foundation-sites but it's still confusing that the one mentioned code part affects the type check.

**Comment by DanielRuf** [comment]
Also imo a deprecation is still not directly a change in the code part but only a warning on the console that a deprecated feature is used.

Here it was one step too much I think. Never saw other solutions introducing deprecations which broke production code but just threw warnings (especially when it comes to node modules).

Or do you see it different?

**Comment by mgol** [comment]
Most of the time it is also a change in the code. We move deprecated APIs to the `deprecated` module which can be excluded in a custom build.

Deprecation is unrelated to this issue, though. We still need a test case.

**Comment by DanielRuf** [comment]
Ok, here are my test results:

The issue comes from the function definition `domManip`.
There we have `scripts = jQuery.map( getAll( fragment, "script" ), disableScript );`
A short test yields that `getAll( fragment, "script" ).length` is `0`.
In `jQuery.map` we use `if ( isArrayLike( elems ) ) {` which contains
```javascript
var length = !!obj && "length" in obj && obj.length,
		type = toType( obj );

	if ( isFunction( obj ) || isWindow( obj ) ) {
		return false;
	}
```

`console.log(type, isFunction( obj ))` results in `object true`.

Reverting it to
```javascript
var length = !!obj && "length" in obj && obj.length,
		type = toType( obj );

	if ( type === "function" || isWindow( obj ) ) {
		return false;
	}
```

fixes it

> We move deprecated APIs to the deprecated module which can be excluded in a custom build.
> 
> Deprecation is unrelated to this issue, though. We still need a test case.

I did not see any change in the `deprecated` module also isn't jQuery.migrate meant for this or did I misinterpret its mission?
> Upgrading libraries such as jQuery can be a lot of work, when breaking changes have been introduced. jQuery Migrate makes this easier, by restoring the APIs that were removed, and additionally shows warnings in the browser console (development version of jQuery Migrate only) when removed and/or deprecated APIs are used.

**Comment by DanielRuf** [comment]
I prepare now a test with chai.js

**Comment by DanielRuf** [comment]
reproducible testcase:
https://github.com/DanielRuf/mocha-jquery-test

try it with `npm install && npm run test`
Then `npm install jquery@3.2.1 && npm run test`

**Comment by DanielRuf** [comment]
cc @ncoden I guess this is the issue. Not sure if another testrunner or testframework or headless browser testing framework would be a solution but many still use mochajs and phantomjs.

**Comment by gibson042** [comment]
Thank you so much for finally providing a test case. The issue is in your [mocha-phantomjs](https://github.com/nathanboktae/mocha-phantomjs) dependency, which itself depends upon an ancient version of phantomjs ([1.9.7-15](https://github.com/nathanboktae/mocha-phantomjs/blob/master/package.json#L42) instead of the [latest 2.1.7](https://www.npmjs.com/package/phantomjs)). That version suffers from a bug in which NodeList objects are functions, thus causing `jQuery.map` to iterate over them as objects rather than as arrays (with the catastrophic results you observe).

For proof, run the following invocations (or npx-free equivalents) in the root directory of your [mocha-jquery-test](https://github.com/DanielRuf/mocha-jquery-test) repository:
* `echo 'console.log("typeof NodeList: " + typeof document.getElementsByTagName("*"))' | npx phantomjs`
* `echo 'console.log("typeof NodeList: " + typeof document.getElementsByTagName("*"))' | npx phantomjs@2.1.7`

We don't support ancient versions of phantomjs, so the correct fix is to get that dependency updated in mocha-phantomjs.

**Comment by DanielRuf** [comment]
Understandable but I see no big reason to for this change as the internal check was the only thing that has changed in the commit.

In this case it should be documented that 3.3.x+ does not support phantomjs 1.x anymore (not sure why this is not caught like in the other parts where jQuery has fixes for phantomjs and some specific browsers).

**Comment by gibson042** [comment]
At this point I'm no longer sure what you're asking for. The `isFunction` behavior change was necessary to fix #3596 and #3600, as [I've already explained](https://github.com/jquery/jquery/pull/3956#discussion_r164288713)—how could we know when a function is _really_ a NodeList? And environments that we fully support are listed at http://jquery.com/browser-support/ , although we could add a note describing [best-effort support for others](https://github.com/jquery/jquery/blob/7eec97aab3476ea0094766dec958e2a0372feac3/Gruntfile.js#L122-L125) ([PRs welcome](https://github.com/jquery/jquery.com/blob/master/pages/browser-support.html)).
